feat: incremental static regeneration

andykenward opened the pull request:
See if we can use incremental static regeneration for next.js pages too improve page load.

We may be able to use this for doc pages too.

https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration

Relates to #33

patrick91 reviewed (approved):
![](https://media.giphy.com/media/Gp32KJtqe08Lu/source.gif)

patrick91 commented:
This is neat, the CoC and Acknowledgements page load super fast now! for the docs I guess it is a bit tricky, since we allow to preview docs, like here: https://github.com/strawberry-graphql/strawberry/pull/687#issuecomment-767645898

maybe we can have a dynamic route just for that use case? 🤔

andykenward commented:
Perhaps we can use [next.js preview mode](
https://nextjs.org/docs/advanced-features/preview-mode) feature for PR?

andykenward commented:
Going to spend some time on the [docs/[[...slug]].ts](https://github.com/strawberry-graphql/strawberry.rocks/blob/499f02ae00bf2ec9771d4c446d022502ecca3590/pages/docs/%5B%5B...slug%5D%5D.tsx#L112-L162) pages to also use incremental static regeneration. 
Need to use [getstaticpaths](https://nextjs.org/docs/basic-features/data-fetching#getstaticpaths-static-generation) to generate all the docs page paths to render.

patrick91 commented:
> Perhaps we can use [next.js preview mode](https://nextjs.org/docs/advanced-features/preview-mode) feature for PR?

that could be an option, but I never tried it, not sure how it works :)

andykenward commented:
@patrick91 Should all the GitHub API queries for doc files be against the latest release tag and not master?
I am planning to allow the website to display different release version docs later on e.g `/docs/tag/1.0.0/general/schema-basics`. But I think the default for `/docs/general/schema-basics` should show the latest release of strawberry.

patrick91 commented:
> @patrick91 Should all the GitHub API queries for doc files be against the latest release tag and not master?

Good point! We can do latest tag :)


> I am planning to allow the website to display different release version docs later on e.g `/docs/tag/1.0.0/general/schema-basics`. But I think the default for `/docs/general/schema-basics` should show the latest release of strawberry.

Sounds perfect 🙌🏼

andykenward commented:
TODO: 
 - pull request support refactor. Going to add a another nextjs page of `/docs/pr/[[...slug]].ts` as the caching and static generation will be different for theses. Having it all in one page is a  bit much.
 - tag/release support for older versions of the docs. This will also have its own nextjs page of `/docs/tag/[[...slug]].ts` as this will be static generated at build and shouldn't need revalidation.

andykenward commented:
The docs should be a lot faster. There is a flash of white when you first hit the docs index page. I think it is a nextjs issue as it doesn't generate the static page for `[[...slug]].html` with the content of the `docs/index.md`. It doesnt execute the getStaticProps function when `params:{}`. The docs say to add `[]` for slug to `getStaticPaths`. Need to look into it more.

andykenward commented:
Refactored the pull request doc preview.

Visiting `/docs/pr/${pull_request_number}` that has a github label of `ok-to-preview` will render on the site. If the pull request is not labelled then 404s.

A pull request doc pages are rendered on the backend on demand and not at build time.

andykenward commented:
Visiting `docs/tag/${tag}` will render on demand the docs for that version. 

But this will use the current design of the site. So older doc markdown files might not render as expect? 

If we decided to change the site structure or design we might have to keep legacy code to support older versions. Or add logic to limit the tag versions the site will support and direct users to the github repo docs folder for that version instead.

patrick91 commented:
> Visiting `docs/tag/${tag}` will render on demand the docs for that version. There is no next.js revalidation of these pages, as they should never need to be rerender after the first request. I expect if the site gets redeployed then they will?

Sounds good :)

> But this will use the current design of the site. So older doc markdown files might not render as expect?

This should be fine. I think we might ignore versions less than v1 in the dropdown list. We have a lot of releases at the moment. I actually wonder if we should only do docs for x.y, instead of x.y.z, what do you think? /cc @BryceBeagle 

> If we decided to change the site structure or design we might have to keep legacy code to support older versions. Or add logic to limit the tag versions the site will support and direct users to the github repo docs folder for that version instead.

That's an option too, but I don't see us using a lot of fancy things in markdown, so I think should be fine :)

patrick91 commented on pages/docs/pr/[[...slug]].tsx:
```diff
@@ -0,0 +1,111 @@
+    /**
+     * Redirect if no pull request number.
+     */
+    // TODO: Could show message to users to submit a pull request instead?
```
don't think we'd need ^^

patrick91 commented on pages/docs/pr/[[...slug]].tsx:
```diff
@@ -0,0 +1,111 @@
+  const pullNumber = getPullRequestNumber(slugs);
+
+  if (pullNumber == null) {
+    /**
+     * 404 if provided pull request number is not valid. Saves a query.
+     */
+    return { notFound: true };
```
nice!

patrick91 commented on lib/api.ts:
```diff
@@ -0,0 +1,283 @@
+      pull.data.labels.find((label) => label.name === "ok-to-preview") !==
+        undefined;
+
+    if (!safeToPreview) {
+      // TODO: we might want to show a message to the users that a pull request needs the label of "ok-to-preview" instead of throwing an error which 404s.
+      throw new Error("not safe to preview");
+    }
```
yes, this could be a good idea, for now the links are only visible when adding the label, so we can keep it like that :)

patrick91 commented on lib/api.ts:
```diff
@@ -0,0 +1,283 @@
+    };
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error("fetchPullRequest", error);
```
shall we set sentry for all these errors?

patrick91 replied:
we can also implement it later 😊

patrick91 commented on lib/api.ts:
```diff
@@ -0,0 +1,283 @@
+ * fetchContributors uses the GitHub REST API for contributors as the GraphQL
+ * API auth token needs push access. @octokit/plugin-paginate-rest gives us
+ * support for paginatated responses. So if we ever get above 100 contributors
+ * this should handle querying more pages for us.
```
💯

patrick91 reviewed (approved):
This is amazing work! Thank you so much ^^

andykenward commented on helpers/image-paths.ts:
```diff
@@ -0,0 +1,21 @@
+
+import { OWNER, REF, REPO } from "~/lib/api";
+
+export const fixImagePathsPlugin = (
```
Going to refactor this to use an object for the functions parameters

andykenward replied:
This is a great feature in vscode! Even implemented the change to the usages.
<img width="305" alt="CleanShot 2021-02-18 at 11 03 26@2x" src="https://user-images.githubusercontent.com/4893048/108348006-28209d00-71d9-11eb-91c9-abbb6aca7262.png">

BryceBeagle commented:
> I actually wonder if we should only do docs for x.y, instead of x.y.z, what do you think? /cc @BryceBeagle

Historically we've had documentation-only changes in releases that only bump the patch version, or no bump at all. This is fine if we're okay with just waiting for the next x.y release for the documentation improvements to make their way to the website.

andykenward commented:
@patrick91 Is there anything more we need to do before this pull request can be merged?

I've made a few more changes since the last review.

- converted all but the contributors GitHub API calls from REST to GraphQL now.
- generate TypeScript typings for GraphQL Queries using GraphQL Code Generator.
- enabled TypeScript strict mode.
- `npm run build` cli now has a `prebuild` that generates the GraphQL typings and runs a type check.
- updated README.md.

patrick91 commented on README.md:
```diff
@@ -6,6 +6,25 @@ to edit the documentation head to the
+**This shows the latest released version and not the master branch docs.**
+#### `https://strawberry.rocks/docs`
+
+Shows the latest [release](https://github.com/strawberry-graphql/strawberry/releases) from [Strawberry Graphql Repo]. These pages are rendered at build time and revalidated after 30 seconds. When a new release of Strawberry Graphql is done then the site should update [Strawberry GraphQL](https://strawberry.rocks) site by itself. Not requiring a re-deployment. 
```
Nice!

patrick91 reviewed (approved):
Looks great! Will merge when I get home later today ☺️
